### DISCUSSION TITLE
Metadata fields for data source columns

### DISCUSSION THREAD

**Discussion by bill-anderson** [devinit/ddw-analyst-ui#440] [Ideas]
For each column in a data source we need:

1. col-hdr - database column name
2. alias - the column name seen by the user
3. description - a full description of the field
4. details - pseudocode and processing notes for all transformations

Is it possible for this to be in a table that one can export, edit and re-import?

**Comment by akmiller01** [comment]
We've already got `col_name` `col_source_name` (identical for the moment) and `col_description.` https://github.com/devinit/ddw-analyst-ui/blob/develop/data_updates/manual/CSV/meta_columns.csv

Would these change frequently enough to justify creating a new export, edit, import feature?

They're displayed on the Data Source Page right now 
![image](https://user-images.githubusercontent.com/2836840/89643159-2e75aa00-d883-11ea-90ab-dd3186433ff2.png)

**Comment by bill-anderson** [comment]
Is the metadata not stored in a table. If not shouldn’t it be? If it is, isn’t csv export and import relatively trivial?

I think most of our tables could do with better metadata. If so, where is the easiest place to work on this?

**Comment by akmiller01** [comment]
It's stored in that CSV, but in the database the table creation/updating is handled by Django (it's considered a core model). Here's what the Postgres table looks like:
![image](https://user-images.githubusercontent.com/2836840/89655740-8cf95300-d898-11ea-9465-94e67a0445a3.png)

And they're editable in the admin interface: 
![image](https://user-images.githubusercontent.com/2836840/89655825-a8645e00-d898-11ea-9f14-199487cf873a.png)

If you want to make amends to the CSV, I could re-upload any changes you make. It also sounds like we'd need to add a new field `details`?

**Comment by bill-anderson** [comment]
I think we do need a details field. The description should contain a simple summary and the details the small print.

Could that form be moved out of the admin so that more people could access it?

**Comment by akmiller01** [comment]
> Could that form be moved out of the admin so that more people could access it?

I would prefer just to give admin access to the folks we think would need to edit it. A typo in the "source_name" of the column description would break the operation of the query builder for those sources.

**Comment by bill-anderson** [comment]
But the people who need to work on the metadata shouldn’t have admin access. Is there a way of protecting some fields after creation?

**Comment by akmiller01** [comment]
Yes, we could make some fields read-only in the admin interface. But before we code a technical solution to this, do you think that people who work on metadata could make their required changed directly to [the CSV](https://github.com/devinit/ddw-analyst-ui/blob/develop/data_updates/manual/CSV/meta_columns.csv)? And we just re-upload the CSV periodically when there are sufficient changes?

**Comment by bill-anderson** [comment]
I thought that is what I suggested in the first place??

**Comment by akmiller01** [comment]
I thought you were asking for a dedicated feature of the website where users could re-upload updated metadata.

Right now the CSV is a part of the Django side of the data, not the DDW side, so the existing import/update form that @edwinmp created would not work for users to export and re-upload the CSV at will.

Changes would need to be made to the CSV, committed to this repository, and then deployed.

**Comment by bill-anderson** [comment]
It doesn’t make sense for database metadata not to be in DDW. Can the form not read and write to a DDW table?

**Comment by akmiller01** [comment]
Yes, the form can read and write to DDW tables. The new feature required here would be reconfiguring the system to read from an un-managed table for all of the column names and definitions.

Everything is in one database. Tables managed by the Django object-relational mapping (ORM) are in the `public` schema. Un-managed tables that contain the DDW data are in the `repo` schema. Django contains a model definition for the columns, so it can natively read/write to that table through the model definition (and admin interface). We would need to create a custom way of modeling the columns outside of the Django standard if we wanted to store the column table in the `repo` schema and treat it like a piece of data itself.

This would be a sizable refactor, but we can do it if we think it would help streamline our processes.

**Comment by bill-anderson** [comment]
This is worth a proper discussion. Personally I think this is very important in the long run, but not perhaps urgent now compared to other priorities.

**Comment by k8hughes** [comment]
@bill-anderson I wonder if this is a discussion that we need to have again so this can be prioritised or we decide that we are not doing it.  I'm going to make it is discussion for now until we can agree what work we want to take forward.
